### PR TITLE
 mclag enhancements as per HLD at Azure/SONIC#596

### PR DESCRIPTION
**Pull request by gitsabari**
What I did
This PR contains the code changes to support click cli changes for mclag enhancements as per HLD Azure/SONiC#596

Why I did it
Implemented as per the MCLAG enhancements HLD: Azure/SONiC#596

How I verified it
Mclag specific VS tests included part of check-in.

Details:
The PR includes the click cli changes as part of MCLAG enhancements
This PR must work with submitted PR's in other sonic repositories which are listed below.
        > Iccpd: https://github.com/Azure/sonic-buildimage/pull/4819 
	> Swss: https://github.com/Azure/sonic-swss/pull/1331                                                                    
	> swss-common: https://github.com/Azure/sonic-swss-common/pull/405 
	> Mclagsyncd: https://github.com/Azure/sonic-swss/pull/1349                                           
	> Mgmt-FW: https://github.com/Azure/sonic-mgmt-framework/pull/59 
		added mclag sonic yang file https://github.com/Azure/sonic-buildimage/pull/7622

### CODE REVIEW

**Review by rathnasabapathyv — APPROVED** [review]

**Comment by jleveque** [comment]
Retest this please

**Comment by tapashdas** [comment]
retest this please

**Comment by gitsabari** [comment]
@gechiang , @lguohan, @rlhui
can you please help review and approve this merge request

**Comment by gitsabari** [comment]
@gechiang , @lguohan, @rlhui
can you please help review and approve this merge request

**Comment by gitsabari** [comment]
retest this please

**Comment by gitsabari** [comment]
retest this please

**Comment by gitsabari** [comment]
@rathnasabapathyv , @prvattem
can you please approve this PR if there are no more questions

**Review by rathnasabapathyv — APPROVED** [review]

File: config/main.py [hunk 1]
```diff
@@ -668,7 +765,9 @@ def _get_disabled_services_list(config_db):
     return disabled_services_list
 
 
-def _stop_services():
+def _
```
Comment (arlakshm) [inline config/main.py:768]: typo ?

**Review by arlakshm — CHANGES_REQUESTED** [review]
Please add unit tests

File: scripts/soft-reboot [hunk 2]
```diff
@@ -0,0 +1,227 @@
+#!/bin/bash
```
Comment (arlakshm) [inline scripts/soft-reboot:1]: why do we need this script ?
This doesnt seems to be related to MC-lag CLI, please create a separate PR for this change

File: config/main.py [hunk 3]
```diff
@@ -4312,6 +4411,246 @@ def delete(ctx):
 
     sflow_tbl['global'].pop('agent_id')
     config_db.set_entry('SFLOW', 'global', sflow_tbl['global'])
+######
```
[the remark this answers is not in the document]
Comment (gitsabari) [reply config/main.py]: done

File: config/main.py [hunk 4]
```diff
@@ -309,6 +315,43 @@ def interface_name_is_valid(config_db, interface_name):
                     return True
     return False
 
+def mclag_domain_id_valid(domain_id):
```
[the remark this answers is not in the document]
Comment (gitsabari) [reply config/main.py]: moved to different file config/mclag.py

File: config/main.py (hunk 1 above)
Comment (gitsabari) [reply config/main.py]: taken care. this was due to mere from master.  updated main.py to master branch and add new mclag file

File: scripts/soft-reboot (hunk 2 above)
Comment (gitsabari) [reply scripts/soft-reboot]: not required. it came in based on merge. updated main.py to be in sync with master. only change from this PR new mclag.py file and config command addition for mclag in main.py

**Comment by arlakshm** [comment]
@gitsabari, the sonic-utilities build is failing, please take a look
**Logs**
https://dev.azure.com/mssonic/build/_build/results?buildId=9712&view=logs&jobId=83516c17-6666-5250-abde-63983ce72a49&j=83516c17-6666-5250-abde-63983ce72a49&t=454e9439-a1f5-599d-fa04-34c4d15aecae

**Comment by haslersn** [comment]
Is this PR ready to merge?

**Comment by gechiang** [comment]
@gitsabari 
Can you share what is the test coverage percentage now? 90%?

**Comment by lguohan** [comment]
no show command?

**Comment by lguohan** [comment]
the unit test only checks the cli return value, in other unit test, it also checks if the config db is correct or not. not sure why this unit test does not validate config db values.

File: config/mclag.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
+#!/usr/sbin/env python
+
+import click
+import swsssdk
```
Comment (lguohan) [inline config/mclag.py:4]: we should use swss-common, no more swsssdk.

**Comment by gitsabari** [comment]
> @gitsabari
> Can you share what is the test coverage percentage now? 90%?

@gechiang current test coverage is 91%

**Comment by gitsabari** [comment]
> no show command?

show commands are part of sonic-yang. klish changes. existing show commands are mclagdctl dump commands, no new additions

File: config/mclag.py (hunk 5 above)
Comment (gitsabari) [reply config/mclag.py]: @lguohan 
i have taken care of this comment. thanks

**Comment by gitsabari** [comment]
retest this please

File: tests/mclag_test.py [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+        key_found = False 
+        for k in keys:
+            if k == domain_id:
+                key_found = True
+        if key_found is False: 
+            return False 
+
+        mclag_entry = db.cfgdb.get_entry("MCLAG_DOMAIN", MCLAG_DOMAIN_ID)
+        if src_ip is not None:
+            temp = mclag_entry["source_ip"]
```
Comment (haslersn) [inline tests/mclag_test.py:73]: Did you mean
```suggestion
            temp = mclag_entry.get("source_ip")
```

File: tests/mclag_test.py [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+            if temp is not None and temp != src_ip:
+                return False
+        if peer_ip is not None:
+            temp = mclag_entry["peer_ip"]
```
Comment (haslersn) [inline tests/mclag_test.py:77]:
```suggestion
            temp = mclag_entry.get("peer_ip")
```

File: tests/mclag_test.py [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+            if temp is not None and temp != peer_ip:
+                return False
+        if peer_link is not None:
+            temp = mclag_entry["peer_link"]
```
Comment (haslersn) [inline tests/mclag_test.py:81]:
```suggestion
            temp = mclag_entry.get("peer_link")
```

Comment (gitsabari) [reply tests/mclag_test.py]: @haslersn thanks for suggestions. taken care

File: tests/mclag_test.py (hunk 7 above)
Comment (gitsabari) [reply tests/mclag_test.py]: @haslersn thanks for suggestions. addressed it

File: tests/mclag_test.py (hunk 6 above)
Comment (gitsabari) [reply tests/mclag_test.py]: thanks for suggestions. addressed it

**Comment by gitsabari** [comment]
> the unit test only checks the cli return value, in other unit test, it also checks if the config db is correct or not. not sure why this unit test does not validate config db values.

@lguohan thanks for the comments. addressed it in latest

**Comment by gitsabari** [comment]
@arlakshm  have addressed all comments and all checks are passed. please help approve and merge the changes

File: config/mclag.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
+#!/usr/sbin/env python
```
Comment (arlakshm) [inline config/mclag.py:1]: do we need this ?

**Review by arlakshm — COMMENTED** [review]
lgtm, one minor comment.

File: config/mclag.py (hunk 9 above)
Comment (gitsabari) [reply config/mclag.py:1]: it is not required. but it shouldn't harm.if required i can remove it

Comment (gitsabari) [reply config/mclag.py]: @arlakshm i have adddressed the comment and resubmitted

**Review by arlakshm — APPROVED** [review]

**Comment by haslersn** [comment]
Can this get merged?

**Pull request merged**